File: hashgraph network/local-network/general/monitoring/monitor_block_sync.js
const { Client, MirrorClient, AccountId, PrivateKey } = require("@hashgraph/sdk");
const winston = require("winston");
const fs = require("fs");
const path = require("path");
const axios = require("axios");
require("dotenv").config();

// Configure logging
const logDir = path.join(__dirname, "../../logs");
if (!fs.existsSync(logDir)) {
    fs.mkdirSync(logDir, { recursive: true });
}

const logger = winston.createLogger({
    level: "info",
    format: winston.format.combine(
        winston.format.timestamp(),
        winston.format.json()
    ),
    transports: [
        new winston.transports.File({ filename: path.join(logDir, "block_sync.log") }),
        new winston.transports.Console({
            format: winston.format.combine(
                winston.format.colorize(),
                winston.format.simple()
            )
        })
    ]
});

// Environment variables
const BLOCK_SYNC_MONITORING_INTERVAL = process.env.BLOCK_SYNC_MONITORING_INTERVAL || 30;
const MINIMUM_INCREMENTAL_NUMBER = process.env.BLOCK_SYNC_MONITORING_MINIMUM_INCREMENTAL_NUMBER || 1;
const HEDERA_NETWORK = process.env.HEDERA_NETWORK || "testnet";
const HEDERA_ACCOUNT_ID = process.env.HEDERA_ACCOUNT_ID;
const HEDERA_PRIVATE_KEY = process.env.HEDERA_PRIVATE_KEY;
const KENYAN_STOCK_API_KEY = process.env.KENYAN_STOCK_API_KEY;
const KENYAN_STOCK_API_ENDPOINT = process.env.KENYAN_STOCK_API_ENDPOINT;

// Initialize Hedera clients
const client = Client.forTestnet();
if (HEDERA_ACCOUNT_ID && HEDERA_PRIVATE_KEY) {
    client.setOperator(HEDERA_ACCOUNT_ID, HEDERA_PRIVATE_KEY);
}

const mirrorClient = new MirrorClient(HEDERA_NETWORK === "mainnet" ? "mainnet" : "testnet");

async function checkHederaSync(startBlockNumber) {
    try {
        // Get latest block from Hedera
        const latestBlock = await mirrorClient.getLatestBlock();
        const latestBlockNumber = latestBlock.number;

        if (latestBlockNumber - startBlockNumber > MINIMUM_INCREMENTAL_NUMBER) {
            logger.info({
                message: "Hedera blocks are successfully synchronized",
                start: startBlockNumber,
                latest: latestBlockNumber
            });
            startBlockNumber = latestBlockNumber;
        } else {
            logger.error({
                message: "FATAL: Hedera block number has not been increased",
                start: startBlockNumber,
                latest: latestBlockNumber
            });
        }

        // Update last block sync timestamp
        const dataDir = path.join(__dirname, "../../data");
        if (!fs.existsSync(dataDir)) {
            fs.mkdirSync(dataDir, { recursive: true });
        }
        fs.writeFileSync(path.join(dataDir, "last_block_sync"), Date.now().toString());

        return startBlockNumber;
    } catch (error) {
        logger.error({
            message: "Error monitoring Hedera sync",
            error: error.message
        });
        return startBlockNumber;
    }
}

async function checkStockMarketSync() {
    try {
        const response = await axios.get(
            `${KENYAN_STOCK_API_ENDPOINT}/market-data`,
            {
                headers: {
                    Authorization: `Bearer ${KENYAN_STOCK_API_KEY}`
                },
                timeout: 10000
            }
        );

        if (response.status === 200) {
            const data = response.data;
            const latestUpdate = new Date(data.lastUpdate);
            const currentTime = new Date();
            
            // Check if data is less than 5 minutes old
            if ((currentTime - latestUpdate) / 1000 < 300) {
                logger.info("Kenya Stock Exchange data is up to date");
                return true;
            } else {
                logger.warning("Kenya Stock Exchange data is stale");
                return false;
            }
        } else {
            logger.error({
                message: "Failed to fetch Kenya Stock Exchange data",
                status: response.status
            });
            return false;
        }
    } catch (error) {
        logger.error({
            message: "Error fetching Kenya Stock Exchange data",
            error: error.message
        });
        return false;
    }
}

async function monitorSync(startBlockNumber) {
    try {
        // Check Hedera sync
        startBlockNumber = await checkHederaSync(startBlockNumber);
        
        // Check Stock Market sync
        await checkStockMarketSync();

        // Check for stock market transactions in the latest block
        const latestBlock = await mirrorClient.getLatestBlock();
        const transactions = await mirrorClient.getBlockTransactions(latestBlock.number);
        
        for (const tx of transactions) {
            if (tx.memo && tx.memo.includes("STOCK_MARKET")) {
                logger.info({
                    message: "Stock market transaction detected",
                    blockNumber: latestBlock.number,
                    transactionId: tx.transaction_id
                });
            }
        }

        return startBlockNumber;
    } catch (error) {
        logger.error({
            message: "Error in monitor sync",
            error: error.message
        });
        return startBlockNumber;
    }
}

async function main() {
    let blockNumber = 0;
    
    while (true) {
        const startTime = Date.now();
        
        blockNumber = await monitorSync(blockNumber);
        
        const elapsedTime = Date.now() - startTime;
        const sleepTime = Math.max(BLOCK_SYNC_MONITORING_INTERVAL * 1000 - elapsedTime, 0);
        await new Promise(resolve => setTimeout(resolve, sleepTime));
    }
}

// Handle process termination
process.on("SIGTERM", () => {
    logger.info("Received SIGTERM signal. Shutting down...");
    process.exit(0);
});

process.on("SIGINT", () => {
    logger.info("Received SIGINT signal. Shutting down...");
    process.exit(0);
});

// Start monitoring
main().catch(error => {
    logger.error({
        message: "Fatal error in main process",
        error: error.message
    });
    process.exit(1);
}); 

File: hashgraph network/hashgraph-network/general/monitoring/monitor_block_sync.py

import logging
import os
import sys
import time
from logging.config import dictConfig
from datetime import datetime

from requests.exceptions import ConnectionError
from web3 import Web3
import requests
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

# Configuration
BLOCK_SYNC_MONITORING_INTERVAL = int(os.environ.get("BLOCK_SYNC_MONITORING_INTERVAL", "30"))
MINIMUM_INCREMENTAL_NUMBER = int(os.environ.get("BLOCK_SYNC_MONITORING_MINIMUM_INCREMENTAL_NUMBER", "1"))
KENYAN_STOCK_API_KEY = os.environ.get("KENYAN_STOCK_API_KEY")
KENYAN_STOCK_API_ENDPOINT = os.environ.get("KENYAN_STOCK_API_ENDPOINT")

# Initialize Web3 with Hedera JSON-RPC endpoint
web3 = Web3(Web3.HTTPProvider(os.environ.get("HEDERA_JSON_RPC_URL", "http://localhost:8545")))

LOG_CONFIG = {
    "version": 1,
    "handlers": {
        "console": {
            "class": "logging.StreamHandler",
            "stream": sys.stdout,
        }
    },
    "loggers": {
        "monitor": {
            "handlers": ["console"],
            "propagate": False,
        }
    },
    "root": {
        "level": "INFO",
    },
}
dictConfig(LOG_CONFIG)
log_fmt = "%(levelname)s [%(asctime)s|MONITOR-HEDERA-STOCK] %(message)s"
logging.basicConfig(format=log_fmt)

def check_hedera_sync(start_block_number):
    """Monitor Hedera network synchronization

    :param start_block_number: Monitoring start block number
    :return: Next monitoring start block number
    """
    try:
        latest_block_number = web3.eth.block_number
        if latest_block_number - start_block_number > MINIMUM_INCREMENTAL_NUMBER:
            logging.info(
                f"Hedera blocks are successfully synchronized: "
                f"start={start_block_number}, "
                f"latest={latest_block_number}"
            )
            start_block_number = latest_block_number
        else:
            logging.error(
                f"FATAL: Hedera block number has not been increased: "
                f"start={start_block_number}, "
                f"latest={latest_block_number}"
            )
    except ConnectionError:
        logging.warning("Unable to connect to Hedera node")
    except Exception as err:
        logging.exception(
            "An exception occurred while monitoring Hedera synchronization: ", err
        )
    finally:
        return start_block_number

def check_stock_market_sync():
    """Monitor Kenya Stock Exchange data synchronization"""
    try:
        headers = {"Authorization": f"Bearer {KENYAN_STOCK_API_KEY}"}
        response = requests.get(
            f"{KENYAN_STOCK_API_ENDPOINT}/market-data",
            headers=headers,
            timeout=10
        )
        
        if response.status_code == 200:
            data = response.json()
            latest_update = datetime.fromisoformat(data.get("lastUpdate"))
            current_time = datetime.utcnow()
            
            # Check if data is less than 5 minutes old
            if (current_time - latest_update).total_seconds() < 300:
                logging.info("Kenya Stock Exchange data is up to date")
                return True
            else:
                logging.warning("Kenya Stock Exchange data is stale")
                return False
        else:
            logging.error(f"Failed to fetch Kenya Stock Exchange data: {response.status_code}")
            return False
            
    except requests.exceptions.RequestException as err:
        logging.exception("Error fetching Kenya Stock Exchange data: ", err)
        return False

def monitor_sync(start_block_number):
    """Monitor both Hedera and Kenya Stock Exchange synchronization

    :param start_block_number: Monitoring start block number
    :return: Next monitoring start block number
    """
    # Check Hedera sync
    start_block_number = check_hedera_sync(start_block_number)
    
    # Check Stock Market sync
    check_stock_market_sync()
    
    return start_block_number

if __name__ == "__main__":
    block_number = 0
    while True:
        start_time = time.time()

        block_number = monitor_sync(start_block_number=block_number)

        elapsed_time = time.time() - start_time
        time.sleep(max(BLOCK_SYNC_MONITORING_INTERVAL - elapsed_time, 0))
